File: php/testcode/fileupload_042.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_upload_s3_no_direct_path
function fileupload042(BenchmarkRequest $req): BenchmarkResponse {
    $file   = $_FILES['f'];
    $key    = 'uploads/' . bin2hex(random_bytes(16));
    $s3     = $req->s3();
    $s3->putObject([
        'Bucket'               => 'app-uploads',
        'Key'                  => $key,
        'SourceFile'           => $file['tmp_name'],
        'ServerSideEncryption' => 'AES256',
    ]); // vuln-code-snippet safe-line php_upload_s3_no_direct_path
    $url = $s3->createPresignedRequest(
        $s3->getCommand('GetObject', ['Bucket' => 'app-uploads', 'Key' => $key]),
        '+1 hour'
    )->getUri();
    return BenchmarkResponse::ok((string) $url);
}
// vuln-code-snippet end php_upload_s3_no_direct_path

File: rust/testcode/fileupload_042.rs
//! CWE-434: Safe upload helper generates server-controlled path, ignoring user-supplied name.

// vuln-code-snippet start testcodeFileupload042
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let _filename = req.param("filename");
    let content = req.param("content");
    let path = safe_upload_path(&_filename);
    let _ = std::fs::write(&path, content.as_bytes()); // vuln-code-snippet target-line testcodeFileupload042
    super::shared::BenchmarkResponse::ok("Uploaded")
}

fn safe_upload_path(_user_name: &str) -> String {
    format!("/uploads/{}.dat", "server-uuid-abc")
}
// vuln-code-snippet end testcodeFileupload042
